tfhe/src/integer/server_key/radix_parallel/scalar_add.rs

```diff
@@ -269,12 +276,560 @@ impl ServerKey {
+        // generates a carry, propagates, or does nothing
+        //
+        // The LUT for the first packing is not the same as other packing,
+        // consequently LUTs for other packing in the first _grouping_ are
+        // not the same as LUTs for other _groupings_.
+        let packed_block_state_luts = {
+            let mut luts = Vec::with_capacity(grouping_size);
+            // We know the first packing is not going to receive any carry
+            // so, we use that to our advantage.
+            luts.push(self.key.generate_lookup_table(|first_block| {
+                let result =
+                    first_block + u64::from(packed_scalar_blocks[0]) + u64::from(input_carry);
+                if result >= packed_modulus {
+                    1 // Generate
+                } else {
+                    0 // Nothing
+                }
+            }));
+
+            // LUTs for other packing in the _first grouping_
+            for (index_in_grouping, packed_scalar_block) in packed_scalar_blocks
+                .get(1..grouping_size.min(packed_scalar_blocks.len()))
+                .unwrap_or(&[])
+                .iter()
+                .copied()
+                .enumerate()
+            {
+                luts.push(
+                    // As the first block in the first grouping is not going to be in propagate
+                    // state, we also use that for other blocks in the grouping
+                    // in order to save a bit, which is going to be important
+                    // later
+                    self.key.generate_lookup_table(|packed_block| {
+                        let result = packed_block + u64::from(packed_scalar_block);
+                        let state = if result >= packed_modulus {
+                            2 // Generate
+                        } else if result == packed_modulus - 1 {
+                            1 // Propagate
+                        } else {
+                            0 // Nothing
+                        };
+
+                        state << index_in_grouping
+                    }),
+                );
+            }
+
+            for (i, packed_scalar_block) in packed_scalar_blocks
+                .get(grouping_size..)
+                .unwrap_or(&[])
+                .iter()
+                .copied()
+                .enumerate()
+            {
+                let index_in_grouping = i % grouping_size;
+                luts.push(self.key.generate_lookup_table(|packed_block| {
+                    let result = packed_block + u64::from(packed_scalar_block);
+                    let state = if result >= packed_modulus {
+                        2 // Generate
+                    } else if result == packed_modulus - 1 {
+                        1 // Propagate
+                    } else {
+                        0 // Nothing
+                    };
+
+                    state << index_in_grouping
```

Comment:
Maybe I'm missing something but it looks like the exact same LUT as the one used in the first group

Reply:
Yes the lut is the same, its the captured value of index_in_grouping that changes
I extrated the logic of this in a local closure

Reply:
But since we are `< grouping_size` in the first group it should yield the same value anyways ?

Reply:
Ah yes indeed that should work, I've changed the way these are written so many times that some things wre not properly changed

Reply:
Well actually no there is still a problem

Reply:
Yes you're right ! I've missed the `.get(1..).enumerate()` which is different than `.enumerate().get(1..)`.

Reply:
Maybe that's worthy of a comment ! To avoid an aggressive refactor by someone going through this code in a few years :)
You currently state that the first group is handled differently but not really how

Reply:
I added one